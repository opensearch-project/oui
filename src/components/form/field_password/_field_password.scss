/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.ouiFieldPassword {
  @include ouiFormControlStyle;
  @include ouiFormControlWithIcon($isIconOptional: false);
  @include ouiFormControlIsLoading;

  &.ouiFieldPassword--compressed {
    @include ouiFormControlWithIcon($isIconOptional: false, $side: 'left', $compressed: true);
  }
}

// Only remove Edge's internal reveal button if we're providing a custom one
.ouiFieldPassword--withToggle::-ms-reveal {
  display: none;
}
