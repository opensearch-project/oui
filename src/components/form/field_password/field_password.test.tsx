/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { requiredProps } from '../../../test/required_props';

import { OuiFieldPassword, OuiFieldPasswordProps } from './field_password';

jest.mock('../validatable_control', () => ({
  OuiValidatableControl: 'oui-validatable-control',
}));

const TYPES: Array<OuiFieldPasswordProps['type']> = [
  'password',
  'text',
  'dual',
];

describe('OuiFieldPassword', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiFieldPassword
        name="elastic"
        id="1"
        placeholder="Placeholder"
        value="1"
        onChange={() => {}}
        {...requiredProps}
      />
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    test('isInvalid is rendered', () => {
      const { container } = render(<OuiFieldPassword isInvalid />);

      expect(container).toMatchSnapshot();
    });

    test('fullWidth is rendered', () => {
      const { container } = render(<OuiFieldPassword fullWidth />);

      expect(container).toMatchSnapshot();
    });

    test('isLoading is rendered', () => {
      const { container } = render(<OuiFieldPassword isLoading />);

      expect(container).toMatchSnapshot();
    });

    test('prepend and append is rendered', () => {
      const { container } = render(
        <OuiFieldPassword prepend="String" append="String" />
      );

      expect(container).toMatchSnapshot();
    });

    test('compressed is rendered', () => {
      const { container } = render(<OuiFieldPassword compressed />);

      expect(container).toMatchSnapshot();
    });

    describe('type', () => {
      TYPES.forEach((type) => {
        test(`${type} is rendered`, () => {
          const { container } = render(<OuiFieldPassword type={type} />);

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('dual', () => {
      test('dualToggleProps is rendered', () => {
        const { container } = render(
          <OuiFieldPassword type="dual" dualToggleProps={requiredProps} />
        );

        expect(container).toMatchSnapshot();
      });

      test('dual type also renders append', () => {
        const { container } = render(
          <OuiFieldPassword
            type="dual"
            append={['String', <span>Span</span>]}
          />
        );

        expect(container).toMatchSnapshot();
      });

      test('dual does not mutate the append array prop', () => {
        const props: OuiFieldPasswordProps = {
          type: 'dual',
          append: ['one', 'two'],
          dualToggleProps: {
            'data-test-subj': 'toggleButton',
          },
        };
        const { container } = render(<OuiFieldPassword {...props} />);

        const toggleButton = screen.getByTestId('toggleButton');
        expect(toggleButton).toBeInTheDocument();

        const eyeIcon = container.querySelector('[data-ouiicon-type="eye"]');
        expect(eyeIcon).toBeInTheDocument();

        fireEvent.click(toggleButton);

        const eyeClosedIcon = container.querySelector(
          '[data-ouiicon-type="eyeClosed"]'
        );
        expect(eyeClosedIcon).toBeInTheDocument();
      });
    });
  });
});
