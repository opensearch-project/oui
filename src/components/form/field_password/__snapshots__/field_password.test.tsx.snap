// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiFieldPassword is rendered 1`] = `
<div
  class="ouiFormControlLayout"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        aria-label="aria-label"
        class="ouiFieldPassword testClass1 testClass2"
        data-test-subj="test subject string"
        id="1"
        name="elastic"
        placeholder="Placeholder"
        spellcheck="false"
        type="password"
        value="1"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props compressed is rendered 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--compressed"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--compressed"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props dual dual type also renders append 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--group"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--inGroup ouiFieldPassword--withToggle"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
  <label
    class="ouiFormLabel ouiFormControlLayout__append"
  >
    String
  </label>
  <span
    class="ouiFormControlLayout__append"
  >
    Span
  </span>
  <button
    aria-label="Show password as plain text. Note: this will visually expose your password on the screen."
    class="ouiButtonIcon ouiButtonIcon--primary ouiButtonIcon--empty ouiButtonIcon--xSmall ouiFormControlLayout__append"
    title="Show password as plain text. Note: this will visually expose your password on the screen."
    type="button"
  >
    <span
      aria-hidden="true"
      class="ouiButtonIcon__icon"
      color="inherit"
      data-ouiicon-type="eye"
    />
  </button>
</div>
`;

exports[`OuiFieldPassword props dual dualToggleProps is rendered 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--group"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--inGroup ouiFieldPassword--withToggle"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
  <button
    aria-label="aria-label"
    class="ouiButtonIcon ouiButtonIcon--primary ouiButtonIcon--empty ouiButtonIcon--xSmall ouiFormControlLayout__append testClass1 testClass2"
    data-test-subj="test subject string"
    title="Show password as plain text. Note: this will visually expose your password on the screen."
    type="button"
  >
    <span
      aria-hidden="true"
      class="ouiButtonIcon__icon"
      color="inherit"
      data-ouiicon-type="eye"
    />
  </button>
</div>
`;

exports[`OuiFieldPassword props fullWidth is rendered 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--fullWidth"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--fullWidth"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props isInvalid is rendered 1`] = `
<div
  class="ouiFormControlLayout"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control
      isinvalid="true"
    >
      <input
        class="ouiFieldPassword"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props isLoading is rendered 1`] = `
<div
  class="ouiFormControlLayout"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword-isLoading"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
    <div
      class="ouiFormControlLayoutIcons ouiFormControlLayoutIcons--right"
    >
      <span
        class="ouiLoadingSpinner ouiLoadingSpinner--medium"
      />
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props prepend and append is rendered 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--group"
>
  <label
    class="ouiFormLabel ouiFormControlLayout__prepend"
  >
    String
  </label>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--inGroup"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
  <label
    class="ouiFormLabel ouiFormControlLayout__append"
  >
    String
  </label>
</div>
`;

exports[`OuiFieldPassword props type dual is rendered 1`] = `
<div
  class="ouiFormControlLayout ouiFormControlLayout--group"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword ouiFieldPassword--inGroup ouiFieldPassword--withToggle"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
  <button
    aria-label="Show password as plain text. Note: this will visually expose your password on the screen."
    class="ouiButtonIcon ouiButtonIcon--primary ouiButtonIcon--empty ouiButtonIcon--xSmall ouiFormControlLayout__append"
    title="Show password as plain text. Note: this will visually expose your password on the screen."
    type="button"
  >
    <span
      aria-hidden="true"
      class="ouiButtonIcon__icon"
      color="inherit"
      data-ouiicon-type="eye"
    />
  </button>
</div>
`;

exports[`OuiFieldPassword props type password is rendered 1`] = `
<div
  class="ouiFormControlLayout"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword"
        spellcheck="false"
        type="password"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFieldPassword props type text is rendered 1`] = `
<div
  class="ouiFormControlLayout"
>
  <div
    class="ouiFormControlLayout__childrenWrapper"
  >
    <oui-validatable-control>
      <input
        class="ouiFieldPassword"
        spellcheck="false"
        type="text"
      />
    </oui-validatable-control>
    <div
      class="ouiFormControlLayoutIcons"
    >
      <span
        class="ouiFormControlLayoutCustomIcon"
      >
        <span
          aria-hidden="true"
          class="ouiFormControlLayoutCustomIcon__icon"
          data-ouiicon-type="lock"
        />
      </span>
    </div>
  </div>
</div>
`;
